File: CODE/leetcodes/src/main/java/com/hit/math/array_string/_283.java
package com.hit.math.array_string;

/**
 * 283. Move Zeroes
 * <p>
 * Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.
 * <p>
 * Example:
 * <p>
 * Input: [0,1,0,3,12]
 * <p>
 * Output: [1,3,12,0,0]
 * <p>
 * Note:
 * <p>
 * You must do this in-place without making a copy of the array.
 * <p>
 * Minimize the total number of operations.
 */
public class _283 {
    public static void moveZeroes(int[] nums) {
        if (nums == null || nums.length == 0) {
            return;
        }
        int insertPos = 0;
        /**
         * Handle all non-zero elements of nums
         */
        for (int num : nums) {
            if (num != 0) {
                nums[insertPos++] = num;
            }
        }
        /**
         * Handle all zero elements of nums
         */
        while (insertPos < nums.length) {
            nums[insertPos++] = 0;
        }
    }

    public static void main(String[] args) {
        int[] nums = {0, 1, 0, 3, 12};
        moveZeroes(nums);
    }
}

File: codes/go/leetcodes/interview/easy_collection/283.go
package easy_collection

/**
 * 283. Move Zeroes
 * <p>
 * Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.
 * <p>
 * Example:
 * <p>
 * Input: [0,1,0,3,12]
 * <p>
 * Output: [1,3,12,0,0]
 * <p>
 * Note:
 * <p>
 * You must do this in-place without making a copy of the array.
 * <p>
 * Minimize the total number of operations.
 */

func moveZeroes(nums []int) []int {
	if len(nums) == 0 {
		return nums
	}
	nonZeroLength := 0
	for _, n := range nums {
		if n != 0 {
			nums[nonZeroLength] = n
			nonZeroLength++
		}
	}
	for ; nonZeroLength < len(nums); nonZeroLength++ {
		nums[nonZeroLength] = 0
	}
	return nums
}
